frame/support/procedural/src/construct_runtime/expand/task.rs

```diff
@@ -0,0 +1,100 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+use crate::construct_runtime::{parse::PalletPath, Pallet};
+use proc_macro2::{Ident, TokenStream};
+use quote::quote;
+
+/// Expands aggregate `RuntimeTask` enum.
+pub fn expand_outer_task(pallet_decls: &[Pallet], scrate: &TokenStream) -> TokenStream {
+	let mut conversion_fns = Vec::new();
+	let mut task_variants = Vec::new();
+	for decl in pallet_decls {
+		if let Some(_) = decl.find_part("Task") {
+			// TODO: for some reason `find_part` above never finds `Task` even when it is
+			// clearly in the pallet
```

Comment:
@KiChjang do you recall when you were setting up `HoldReason` whether you ever ran into this issue? I based this `expand_outer_task` method off of your `expand_outer_hold_reason` method almost exactly and tried to follow all the conventions your `HoldReason` follows as far as I can tell.

Your version is here: https://github.com/paritytech/substrate/blob/34f45c6e7cdd590d4f29c73827dde796a8b03a1c/frame/support/procedural/src/construct_runtime/expand/hold_reason.rs#L22-L51

What's happening here is for some reason `decl.find_part("Task")` is coming back empty, even when we are looking at a pallet like `TasksExample` which 100% does have a `pub enum Task`.

Is there anything additional I need to do to make this get picked up by `find_part`?

I'll be looking deeply into how pallet parts are parsed and constructed next to see why this might be going around, but wanted to double check there isn't something obvious I've missed first.

Reply:
Yes, you need to an an additional variant to one of the enums in the parse module of construct_runtime.

Reply:
So I already am on this one:

https://github.com/paritytech/substrate/pull/14329/files#diff-bc4d1d4294aa1ace0173ab00f6fac84463c64992cd3857099a000fb2266388cfR497

Is there some other I'm missing?

Reply:
This is the enum I'm talking about:

https://github.com/paritytech/substrate/blob/34f45c6e7cdd590d4f29c73827dde796a8b03a1c/frame/support/procedural/src/construct_runtime/parse.rs#L395-L409

In addition to modifying the above, the _runtime maintainer_ (i.e. whoever calls `construct_runtime`) needs to also import the pallet part in the `construct_runtime` macro, e.g. 

```rust
construct_runtime! {
pub enum Runtime
where
    // ...
    {
        System: frame_system,
        MyPallet: path::to::my::pallet::{Pallet, Storage, Call, Event<T>, Task},
    }
}
```

Note the additional `Task` keyword used in `MyPallet`. You can also use the implicit pallet part import syntax so you have less hassle (i.e. simply `MyPallet: path::to::my::pallet,`).

Reply:
So I believe I am already doing both of these...

In `PalletPartKeyword`:
https://github.com/paritytech/substrate/pull/14329/files#diff-bc4d1d4294aa1ace0173ab00f6fac84463c64992cd3857099a000fb2266388cfR408

In `construct_runtime!`:
https://github.com/paritytech/substrate/pull/14329/files#diff-67684005af418e25ff88c2ae5b520f0c040371f1d817e03a3652e76b9485224aR1956

Perhaps I'm not in the right `construct_runtime!`?? @KiChjang

Reply:
ahhh OK I followed your syntax and now I am just getting this error:

```
  error[E0412]: cannot find type `Task` in crate `tasks_example`
      --> /home/sam/workspace/substrate/bin/node/runtime/src/lib.rs:1883:1
       |
  652  |  |     #[compact]
       |  |_________^ not found in `tasks_example`
  ...
  1883 | // construct_runtime!(
  1884 | |      pub struct Runtime
  1885 | |      {
  1886 | |          System: frame_system,
  ...    |
  1957 | |      }
  1958 | |  );
       | |__- in this macro invocation
       |
       = note: this error originates in the macro `frame_support::construct_runtime` which comes from the expansion of the macro `construct_runtime` (in Nightly builds, run with -Z macro-backtrace for more info)
  help: consider importing one of these items
       |
  25   + use crate::tasks_example::pallet::Task;
       |
  25   + use frame_support::pallet_prelude::Task;
       |
  25   + use tasks_example::pallet::Task;
       |

  For more information about this error, try `rustc --explain E0412`.
  ```
  
  Strangely, `tasks_example::Task` is perfectly accessible according to the language server, and I can see in the file that there is indeed `pub use pallet::*;` so yeah investigating this one now

Reply:
So yeah this is super weird. I went ahead and pushed up the current code. Anyone have any idea how/why this error is happening when that type is clearly accessible from that context?

Reply:
@sam0x17 Check out tt_default_parts macro. That adds the correct parts during construct_runtime expansion. You would probably need something similar for Task.